### PR TITLE
Fix auto inventory letters bugs #45625, #45401, #43254

### LINKED ISSUE #43254: Auto-inventory letters duplicating
# Describe the bug

My auto-inventory letters are set to "favorites" in the options menu.

I am carrying some aspirin that is favorited as "a", I pick up another 20 units on top of the 19 I had, immediately after doing that another item in my inventory (a teleporter) that had no letter assigned to it, then becomes "a" as well.

# Steps To Reproduce

I've included a save file; The character in the save is right next to some aspirin, pick it up and the teleporter that is in that character's inventory will duplicate the "a" from the aspirin. The save file uses the following mods: (remove dildoes is just an item exclusion mod and I don't think it will matter, regardless, I'll include it if you want)

[Save.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5134518/Save.zip)
[remove_dildoes.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5134529/remove_dildoes.zip)

Untitled

# Expected behavior
For it to not duplicate letters

# Screenshots
Untitled

# Versions and configuration
Windows 10
0.E-5407-g8b385a2 (tiles), build 10952

# Additional context
Note: I did use the debug menu to edit some mutations for my character, but no other changes.

Superfluous trivia: This has happened with other characters in the past few months or so, but none bothered me enough to make a bug report like this. You might have guessed I lost some progress as a result of the teleporter having the same keybind as my aspirin.

### LINKED ISSUE #45401: Inventory - Item hotkeys assigned only counts worn/wielded items
# Describe the bug

When viewing the `i`nventory, the "Item hotkeys assigned" indicator in the window's header only counts hotkey-items that are worn or wielded, not any other items in the inventory.

# Steps To Reproduce

1. Go to `i`nventory screen
2. Select an item and press enter
3. Press `=`, and assign a hotkey to the item
4. Re-open the inventory screen - hotkey is only counted if the item is wielded or worn

# Expected behavior

All items with hotkeys assigned should be included in the count. If I have 5 items with hotkeys, the inventory header should say "Item hotkeys assigned: 5/72" regardless of whether I am wearing, wielding, or storing the item.

# Screenshots

Here I have 5 items with hotkeys in my inventory (`c`ellphone, `m`atchbook, `g`ame watch, combat `k`nife, and `h`eadlamp), but "Item hotkeys assigned" only shows "2/72", since I am only wielding/wearing two of them:

image

If I also wear the game watch, it is counted (but the `c`ellphone and `m`atchbook are not):

image

And if I take off all items and put them in my backpack, it shows "0/72" item hotkeys are assigned:

image

# Versions and configuration

- OS: Linux
    - OS Version: Distributor ID: Pop; Description: Pop!_OS 20.04 LTS; Release: 20.04; Codename: focal; 
- Game Version: 0.E-7515-gbcdd3a2561 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions]
]

### LINKED ISSUE #45625: "Auto inventory letters" feature does not working.
# Describe the bug
When an item is set to favorited, no hotkey is assigned to the item.

# Steps To Reproduce
1. Set "Auto inventory letters" in Option > Interface, to "Favorites".
2. Open inventory and highlight any item in inventory.
3. Use hot key "*" to assign that item as favorited.

# Expected behavior
When assign an item as favorited, a hotkey in alphabet order should automatic be assigned to the favorited item.

# Versions and configuration
 - OS: Windows 7
 - Game Version: 0.E-7497-g48b6f14 [64-bit]
 - Graphics version: Tiles
 - Ingame language: English
 - Mods loaded: dda, no_npc_food, package_bionic_professions, magiclysm, crazy_cataclysm, my_sweet_cataclysm, Graphical_Overmap, Graphical_Overmap_Magiclysm, speedydex, stats_through_kills, StatsThroughSkills.

### PR DESCRIPTION
**Pull request by exuvo**
#### Summary
Bugfixes "Fix auto inventory letters not working with pockets"

#### Purpose of change
Fix auto inventory letters not assigning letters to items that are picked up into pockets. Only pickup directly to wield or wear was working.
Fixes #45625
Fixes #45401

Fix auto inventory letters not counting used labels from items pockets. Counter at the top was ignoring non wield/wear items with letters. Could duplicate letters as newly wielded item could get letter already used by item inside a pocket.
Fixes #43254

#### Describe the solution
When picking up into a pocket call inv->update_invlet unless item got merged into an existing stack.

Use character.visit_items instead of just weapon and worn.

#### Testing
Wield, wear or pickup items from the ground. Make sure all get letters assigned.
Check that used letters counter at the top of inventory is correct.

#### Additional context
Liquids inside containers does not get a letter assigned. Not sure if needed as the container does get a letter already. Might use up too much letters if both get a letter.
Ex pickup a soda can. Can gets letter but not the extra row for the soda.

Also how the fuck did you manage to release F with this feature broken. Are we really so few using this feature?

### CODE REVIEW

**Comment by exuvo** [comment]
Also tested applying to 0.F-branch which is what i will be playing on.

File: src/character.cpp [hunk 1]
```diff
@@ -2906,6 +2906,11 @@ item *Character::try_add( item it, const item *avoid, const bool allow_wield )
     } else {
         // this will set ret to either it, or to stack where it was placed
         pocket.second->add( it, &ret );
+        
+        if (!keep_invlet && (!it.count_by_charges() || (it.charges == ret->charges))) {
```
Comment (kevingranade) [inline src/character.cpp:2910]:
```suggestion
        if( !keep_invlet && ( !it.count_by_charges() || it.charges == ret->charges ) ) {
```

File: src/character.cpp [hunk 2]
```diff
@@ -2906,6 +2906,11 @@ item *Character::try_add( item it, const item *avoid, const bool allow_wield )
...
+            inv->update_invlet( *ret );
+        }
+
```
Comment (kevingranade) [inline src/character.cpp:2913]:
```suggestion

```

File: src/character.cpp [hunk 3]
```diff
@@ -3316,10 +3321,10 @@ invlets_bitset Character::allocated_invlets() const
 {
     invlets_bitset invlets = inv->allocated_invlets();
 
-    invlets.set( weapon.invlet );
-    for( const auto &w : worn ) {
-        invlets.set( w.invlet );
-    }
+    visit_items([&invlets] (item * i, item *) -> VisitResponse {
+        invlets.set(i->invlet);
```
Comment (kevingranade) [inline src/character.cpp:3325]:
```suggestion
        invlets.set( i->invlet );
```

File: src/character.cpp [hunk 4]
```diff
@@ -2906,6 +2906,11 @@ item *Character::try_add( item it, const item *avoid, const bool allow_wield )
     } else {
         // this will set ret to either it, or to stack where it was placed
         pocket.second->add( it, &ret );
+
```
Comment (kevingranade) [inline src/character.cpp:2909]:
```suggestion

```

File: src/character.cpp [hunk 5]
```diff
@@ -3316,10 +3321,10 @@ invlets_bitset Character::allocated_invlets() const
 {
     invlets_bitset invlets = inv->allocated_invlets();
 
-    invlets.set( weapon.invlet );
-    for( const auto &w : worn ) {
-        invlets.set( w.invlet );
-    }
+    visit_items([&invlets] (item * i, item *) -> VisitResponse {
```
Comment (kevingranade) [inline src/character.cpp:3324]:
```suggestion
    visit_items( [&invlets] ( item * i, item * ) -> VisitResponse {
```

File: src/character.cpp [hunk 6]
```diff
@@ -3316,10 +3321,10 @@ invlets_bitset Character::allocated_invlets() const
...
+        invlets.set(i->invlet);
+        return VisitResponse::NEXT;
+    });
```
Comment (kevingranade) [inline src/character.cpp:3327]:
```suggestion
    } );
```

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/auto-inventory-letters-not-working-properly-or-at-all-in-0-f/26717/2

**Pull request merged**